### PR TITLE
[MAINTENANCE] Modify cloud-based `DataConnectors` attrs to fit existing `TupleStoreBackend` API

### PR DESCRIPTION
**Pull request by cdkini**
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- Refactor all instances of GCS `bucket_or_name` to `bucket` to align with the existing `TupleGCSStoreBackend` API
- Refactor all instances of Azure `name_starts_with` to `prefix` to align with the existing `TupleAzureStoreBackend` API
- Change GCS `max_results` to existing `max_keys` used in `Asset` as I deemed them somewhat equivalent and felt that keeping both attrs was a bit duplicative.

Since we take a number of these attributes, bundle them into a dictionary, and pass that dictionary as **kwargs to the S3/Azure/GCS SDK, we need to ensure that our argument names are consistent with what is noted in the API. In order to manage this, we need to manually rename these values in our methods.

Since this transformation is abstracted away from the user and results in a clean, uniform API for all `DataConnectors`, I believe that this is the way to go for our initial release of the GCS and Azure features.

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/datasource/data_connector/inferred_asset_gcs_data_connector.py
Comment (cdkini) [inline great_expectations/datasource/data_connector/inferred_asset_gcs_data_connector.py:137]: Since we are doing manual reassignment of the user-provided arguments to different names, I found it important to explicitly point out the API that forces us to do this transformation

File: great_expectations/datasource/data_connector/inferred_asset_gcs_data_connector.py
Comment (cdkini) [inline great_expectations/datasource/data_connector/inferred_asset_gcs_data_connector.py:39]: The portion about using specific terms in order to maintain consistency with the GCS API is no longer relevant. We still adhere to the constraints but do so without the user explicitly needing to know.

File: great_expectations/data_context/types/base.py
Comment (cdkini) [inline great_expectations/data_context/types/base.py]: This was entirely missing from our schema validation

File: great_expectations/datasource/data_connector/inferred_asset_gcs_data_connector.py
Comment (Shinnnyshinshin) [reply great_expectations/datasource/data_connector/inferred_asset_gcs_data_connector.py:137]: Thank you for this explanation 🙇🏼

File: great_expectations/data_context/types/base.py
Comment (Shinnnyshinshin) [reply great_expectations/data_context/types/base.py]: Thank you ❤️

**Review by Shinnnyshinshin — APPROVED** [review]
The work looks great :) thank you for being so thorough in your implementation. Since there are some open questions on #3321 I would like to request that we wait to merge this until that PR has been approved.

**Review by alexsherstinsky — APPROVED** [review]
LGTM as long as:
* Users can read our documentation and learn how to translate the cloud-provider-specific entity names into our common names; and
* The PR https://github.com/great-expectations/great_expectations/pull/3336 is merged first.

Thank you!

**Pull request closed without merging**
